Web Inspector: Elements Tab: create an Event badge for nodes that have event listeners

dcrousso opened the pull request:
#### 2a8187b1de2df352158fefedcf033f80d9cf74b3
<pre>
Web Inspector: Elements Tab: create an Event badge for nodes that have event listeners
<a href="https://bugs.webkit.org/show_bug.cgi?id=244124">https://bugs.webkit.org/show_bug.cgi?id=244124</a>

Reviewed by Patrick Angle.

Only show &quot;Event&quot; badges for nodes that directly have event listeners (i.e. no ancestors). This will
give developers a visual way of seeing where an event could be handled in a particular ancestor path
as they&apos;d be able to scan for &quot;Event&quot; badges to find every ancestor that has an event listener.

Clicking the &quot;Event&quot; badge will show a popover containing details about the attached event listeners,
such as capturing/bubbling, once, enabled/disabled, and breakpoint, all grouped by event type.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
Add an `&quot;event&quot;` enum value to `CSS.LayoutFlag`.

* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didAddEventListenerImpl):
(WebCore::InspectorInstrumentation::willRemoveEventListenerImpl):
Also inform the `InspectorCSSAgent` that an event listener was added/removed.

* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::reset):
(WebCore::layoutFlagContextType):
(WebCore::layoutFlagsContainLayoutContextType): Added.
(WebCore::hasJSEventListener):
(WebCore::InspectorCSSAgent::layoutFlagsForNode):
(WebCore::toProtocol): Added.
(WebCore::InspectorCSSAgent::protocolLayoutFlagsForNode): Added.
(WebCore::pushChildrenNodesToFrontendIfLayoutFlagIsRelevant):
(WebCore::InspectorCSSAgent::didChangeRendererForDOMNode):
(WebCore::InspectorCSSAgent::didAddEventListener): Added.
(WebCore::InspectorCSSAgent::willRemoveEventListener): Added.
(WebCore::InspectorCSSAgent::nodeHasLayoutFlagsChange): Added.
(WebCore::InspectorCSSAgent::nodesWithPendingLayoutFlagsChangeDispatchTimerFired):
(WebCore::layoutFlagContextTypeForRenderer): Deleted.
Keep track of the layout flags for each `Node` to avoid re-dispatching the same flags to the
frontend. This is especially important for event listeners as they can be very noisy (211169@main).

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getEventListenersForNode):
(WebCore::InspectorDOMAgent::buildObjectForNode):
Allow `DOM.getEventListenersForNode` to control whether ancestor event listeners are included.

* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.set layoutFlags):
(WI.DOMNode.prototype.getEventListeners):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype._createBadge):
(WI.DOMTreeElement.prototype._createBadges):
(WI.DOMTreeElement.prototype.async _handleEventBadgeClicked): Added.
(WI.DOMTreeElement.prototype._handleBadgeDoubleClicked): Renamed from `_layoutBadgeDoubleClicked`.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.css:
(.event-badge-popover-content): Added.
(.event-badge-popover-content &gt; .details-section.event-listeners): Added.
(.event-badge-popover-content &gt; .details-section.event-listeners .details-section): Added.
Add logic for creating an &quot;Event&quot; badge that shows the list of event listeners directly attached to
that node (i.e. no ancestors).

* Source/WebInspectorUI/UserInterface/Views/EventListenerSectionGroup.js:
(WI.EventListenerSectionGroup.groupIntoSectionsByEvent): Added.
(WI.EventListenerSectionGroup.groupIntoSectionsByTarget): Added.
(WI.EventListenerSectionGroup._createEventListenerSection): Added.
* Source/WebInspectorUI/UserInterface/Views/EventListenerSectionGroup.css:
(.event-listener-section &gt; .header &gt; .event-listener-options): Added.
(.event-listener-section:hover &gt; .header &gt; .event-listener-options): Added.
* Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
(WI.DOMNodeDetailsSidebarPanel.prototype.async _refreshEventListeners):
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.createEventListenerSection): Deleted.
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.generateGroupsByEvent): Deleted.
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.generateGroupsByTarget): Deleted.
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.eventListenersCallback): Deleted.
* Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.css:
(.sidebar &gt; .panel.dom-node-details .details-section.dom-node-event-listeners .details-section.event-listener-section &gt; .header &gt; .event-listener-options): Deleted.
(.sidebar &gt; .panel.dom-node-details .details-section.dom-node-event-listeners .details-section.event-listener-section:hover &gt; .header &gt; .event-listener-options): Deleted.
Move the code that groups `WI.EventListenerSectionGroup` to be `static` methods on `WI.EventListenerSectionGroup`
so that it can be reused in other places (e.g. inside the `WI.Popover` for the &quot;Event&quot; badge).

* Source/WebInspectorUI/UserInterface/Views/Popover.js:
(WI.Popover.prototype.present):
Add an option to also update the sizing of the `WI.Popover` when re-presenting the same content.
This is necessary because the `WI.Popover` for the &quot;Event&quot; badge is limited in height to the size of
Web Inspector, so resizing Web Inspector should also resize the popover (not just reposition).

* Source/WebInspectorUI/UserInterface/Views/DetailsSection.js:
(WI.DetailsSection):
* Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
(.details-section .details-section &gt; .header):
(.details-section:has(&gt; .header &gt; .title:not(:empty)) .details-section &gt; .header): Added.
Don&apos;t lower the header of a subsection unless the parent section has a `title`. This is necessary
because the `WI.Popover` for the &quot;Event&quot; badge has a top-level `WI.DetailsSection` with no `title`
in order to get the same nested styling applied to the Event Listeners section in the Node panel of
the details sidebar in the Elements Tab.

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-attribute.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-attribute-expected.txt: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-script.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-script-expected.txt: Added.
* LayoutTests/inspector/dom/getEventListenersForNode.html:
* LayoutTests/inspector/dom/getEventListenersForNode-expected.txt:

Canonical link: <a href="https://commits.webkit.org/253727@main">https://commits.webkit.org/253727@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3dc1efc708c7759580c677d473900acfa23c008

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17770 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95728 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149485 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90878 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29344 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79044 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90951 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92515 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23693 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73756 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23714 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78688 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/78975 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66719 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78808 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27095 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12827 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72468 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27030 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13841 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25842 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2645 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28709 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36704 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75253 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28653 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33120 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16644 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -274,7 +274,8 @@
                 "flex",
-                "grid"
+                "grid",
+                "event"
```
Oh no... an event layout flag? Is this the only way we can get this state from the backend to the frontend? I understand the convenience, but having this value in an enum in the CSS domain feels very contrived since this has absolutely nothing to do with layout.

dcrousso replied:
I agree that it doesn't really have anything to do with the idea of "layout" (tho there are things like "event regions" that one could argue as being related to "layout").  I wanted to move this entire thing to be something like `DOM.NodeStateFlag` but that felt a bit overkill for this (not to mention it would involve changing a whole bunch more than just the name of the enum since we'd also probably want to move the events/commands to `DOM` as well).

I don't really feel strongly any way, so I'm happy to do whatever.  I went with this only because it was the least modification.

patrickangle replied:
Hmmm... this is probably fine then.

patrickangle commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -936,44 +938,92 @@ Protocol::ErrorStringOr<void> InspectorCSSAgent::forcePseudoState(Protocol::DOM:
 }
 
-RefPtr<JSON::ArrayOf<String /* Protocol::CSS::LayoutFlag */>> InspectorCSSAgent::layoutFlagsForNode(Node& node)
+static std::optional<InspectorCSSAgent::LayoutFlag> layoutFlagContextType(OptionSet<InspectorCSSAgent::LayoutFlag> layoutFlags)
```
This method and the one above share the same name, despite performing different tasks. Plus, this method is like a second mapping of internal -> protocol values, which we already have in `toProtocol`. Since the result of this method is only ever used for a boolean check, maybe it should be called `layoutFlagsContainLayoutContextType` and have it return a boolean?

patrickangle commented on Source/WebCore/inspector/agents/InspectorCSSAgent.h:
```diff
@@ -123,11 +124,22 @@ class InspectorCSSAgent final : public InspectorAgentBase , public Inspector::CS
     void didModifyDOMAttr(Element&);
 
+    enum class LayoutFlag : uint8_t {
+        Rendered = 1 << 0,
+        Flex = 1 << 1,
+        Grid = 1 << 2,
+        Event = 1 << 3,
```
Is there a compelling reason for us to explicitly define the values backing these flags? A quick search of some other agents tells me we haven't in the past.

dcrousso replied:
this is an optimization to avoid issues on webpages that frequently modify event listeners
> Keep track of the layout flags for each `Node` to avoid re-dispatching the same flags to the frontend. This is especially important for event listeners as they can be very noisy (211169@main).

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -273,10 +273,8 @@ WI.DOMNode = class DOMNode extends WI.Object
-
-        if (Array.shallowEqual(layoutFlags, this._layoutFlags))
-            return;
+        console.assert(layoutFlags.filter((layoutFlag) => WI.DOMNode._LayoutContextTypes.includes(layoutFlag)).length <= 1, layoutFlags);
```
Was changing this assertion necessary, and if so should the assertion on the next line also not use `this._layoutFlags`?

dcrousso replied:
yes because this `console.assert` is about checking the incoming `layoutFlags`, not the previous value of `this._layoutFlags`

the next `console.assert` is about making sure that this incoming `layoutFlags` and previous `this._layoutFlags` are not equal, so both are useful to have (tho yes technically we could drop the `this._layoutFlags` since the only way we'd log the value is if they are equal, so no need to log the value twice)

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2033,14 +2033,19 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
             break;
+
+        case WI.DOMTreeElement.BadgeType.Event:
+            text = WI.UIString("Event");
```
This probably deserves a localization comment since it is about the web platform concept of an Event, not a generic concept.

patrickangle replied:
Also it looks like you haven't updated the strings file in this patch yet.

dcrousso replied:
this is using an existing UI string, which is why it's not in the localization file (and why i didn't add a comment for it, since we'd already have a translation for it)

patrickangle reviewed (commented):
Is there anything we can do about the UI you show in your screenshot. It would be nice if the entire background matched the popover's background (except for maybe the header control)? It looks very strange IMO for there to be a slightly different color box inset within the popover.

dcrousso committed:
Web Inspector: Elements Tab: create an Event badge for nodes that have event listeners
https://bugs.webkit.org/show_bug.cgi?id=244124

Reviewed by Patrick Angle.

Only show "Event" badges for nodes that directly have event listeners (i.e. no ancestors). This will
give developers a visual way of seeing where an event could be handled in a particular ancestor path
as they'd be able to scan for "Event" badges to find every ancestor that has an event listener.

Clicking the "Event" badge will show a popover containing details about the attached event listeners,
such as capturing/bubbling, once, enabled/disabled, and breakpoint, all grouped by event type.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
Add an `"event"` enum value to `CSS.LayoutFlag`.

* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didAddEventListenerImpl):
(WebCore::InspectorInstrumentation::willRemoveEventListenerImpl):
Also inform the `InspectorCSSAgent` that an event listener was added/removed.

* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::reset):
(WebCore::layoutFlagContextType):
(WebCore::layoutFlagsContainLayoutContextType): Added.
(WebCore::hasJSEventListener):
(WebCore::InspectorCSSAgent::layoutFlagsForNode):
(WebCore::toProtocol): Added.
(WebCore::InspectorCSSAgent::protocolLayoutFlagsForNode): Added.
(WebCore::pushChildrenNodesToFrontendIfLayoutFlagIsRelevant):
(WebCore::InspectorCSSAgent::didChangeRendererForDOMNode):
(WebCore::InspectorCSSAgent::didAddEventListener): Added.
(WebCore::InspectorCSSAgent::willRemoveEventListener): Added.
(WebCore::InspectorCSSAgent::nodeHasLayoutFlagsChange): Added.
(WebCore::InspectorCSSAgent::nodesWithPendingLayoutFlagsChangeDispatchTimerFired):
(WebCore::layoutFlagContextTypeForRenderer): Deleted.
Keep track of the layout flags for each `Node` to avoid re-dispatching the same flags to the
frontend. This is especially important for event listeners as they can be very noisy (211169@main).

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getEventListenersForNode):
(WebCore::InspectorDOMAgent::buildObjectForNode):
Allow `DOM.getEventListenersForNode` to control whether ancestor event listeners are included.

* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.set layoutFlags):
(WI.DOMNode.prototype.getEventListeners):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView.prototype._populateConfigureDOMTreeBadgesNavigationItemContextMenu):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype._createBadge):
(WI.DOMTreeElement.prototype._createBadges):
(WI.DOMTreeElement.prototype.async _handleEventBadgeClicked): Added.
(WI.DOMTreeElement.prototype._handleBadgeDoubleClicked): Renamed from `_layoutBadgeDoubleClicked`.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.css:
(.event-badge-popover-content): Added.
(.event-badge-popover-content > .details-section.event-listeners): Added.
(.event-badge-popover-content > .details-section.event-listeners .details-section): Added.
Add logic for creating an "Event" badge that shows the list of event listeners directly attached to
that node (i.e. no ancestors).

* Source/WebInspectorUI/UserInterface/Views/EventListenerSectionGroup.js:
(WI.EventListenerSectionGroup.groupIntoSectionsByEvent): Added.
(WI.EventListenerSectionGroup.groupIntoSectionsByTarget): Added.
(WI.EventListenerSectionGroup._createEventListenerSection): Added.
* Source/WebInspectorUI/UserInterface/Views/EventListenerSectionGroup.css:
(.event-listener-section > .header > .event-listener-options): Added.
(.event-listener-section:hover > .header > .event-listener-options): Added.
* Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.js:
(WI.DOMNodeDetailsSidebarPanel.prototype.async _refreshEventListeners):
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.createEventListenerSection): Deleted.
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.generateGroupsByEvent): Deleted.
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.generateGroupsByTarget): Deleted.
(WI.DOMNodeDetailsSidebarPanel.prototype._refreshEventListeners.eventListenersCallback): Deleted.
* Source/WebInspectorUI/UserInterface/Views/DOMNodeDetailsSidebarPanel.css:
(.sidebar > .panel.dom-node-details .details-section.dom-node-event-listeners .details-section.event-listener-section > .header > .event-listener-options): Deleted.
(.sidebar > .panel.dom-node-details .details-section.dom-node-event-listeners .details-section.event-listener-section:hover > .header > .event-listener-options): Deleted.
Move the code that groups `WI.EventListenerSectionGroup` to be `static` methods on `WI.EventListenerSectionGroup`
so that it can be reused in other places (e.g. inside the `WI.Popover` for the "Event" badge).

* Source/WebInspectorUI/UserInterface/Views/Popover.js:
(WI.Popover.prototype.present):
Add an option to also update the sizing of the `WI.Popover` when re-presenting the same content.
This is necessary because the `WI.Popover` for the "Event" badge is limited in height to the size of
Web Inspector, so resizing Web Inspector should also resize the popover (not just reposition).

* Source/WebInspectorUI/UserInterface/Views/DetailsSection.js:
(WI.DetailsSection):
* Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
(.details-section .details-section > .header):
(.details-section:has(> .header > .title:not(:empty)) .details-section > .header): Added.
Don't lower the header of a subsection unless the parent section has a `title`. This is necessary
because the `WI.Popover` for the "Event" badge has a top-level `WI.DetailsSection` with no `title`
in order to get the same nested styling applied to the Event Listeners section in the Node panel of
the details sidebar in the Elements Tab.

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-attribute.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-attribute-expected.txt: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-script.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Event-script-expected.txt: Added.
* LayoutTests/inspector/dom/getEventListenersForNode.html:
* LayoutTests/inspector/dom/getEventListenersForNode-expected.txt:

Canonical link: https://commits.webkit.org/253727@main
